### ISSUE TITLE
xdist: boxed send an item object to a subprocess

### ISSUE THREAD

**Issue by fontealpina** [pytest-dev/pytest#1137]
I'm aware that currently xdist boxed plugin doesn't run on Windows, due to it is based on linux fork.
But in my project I need a boxed feature in a cross-platform way, with also a control on subprocess memory consumption and time duration (but is a detail here). So I've written a new boxed plugin, here the core code even if in the real code there are more details like subprocess control:  

     from _pytest.runner import runtestprotocol

     def pytest_addoption(parser):
          group = parser.getgroup("boxed", "")
          group._addoption('--report-file', dest="pickle_file_name", action="store",
                           type="string", help="Specify the output file name on which "
                           "the 'reports' pytest object will be saved.")

     def pytest_runtest_protocol(item):
           pickle_file_name = item.config.getvalue("pickle_file_name")

           if pickle_file_name:
               # This is the child code called invoking pytest specifing
               # the option --report-file=pickle_file_name
               reports = runtestprotocol(item, log=False)
               with open(pickle_file_name, "wb") as pkl_file:
                   Pickle.dump(reports, pkl_file)

            args = ['pytest.py', '--report-file=pickled_report', item.nodeid]
            process = subprocess.Popen(args)
            process.wait()

            with open(pickle_file_name, "rb") as pkl_file:
                reports = Pickle.load(pkl_file)

            for rep in reports:
                 item.ihook.pytest_runtest_logreport(report=rep)
            return True  

Basically the idea is to re-run pytest.py in the subprocess with only one test name passed as argument and the filename where the subprocess test report will be pickled. It works but, it is a bit slow primarily due to for every single test in the subprocess the whole pytest needs to be set up. On my machine it takes about 0.3 sec and if there are thousands of tests means hours! Thus to make it faster my idea is maybe to pass directly the item object to subprocess and run just the function

    reports = runtestprotocol(item, log=False)

avoiding to run the whole pytest. But the item object is not pickable.
So how can I pass an item object to a subprocess? Or create it in a faster way into subprocess?

**Comment by RonnyPfannschmidt** [comment]
that approach is next to impossible,

future xdist will include a mode that will have each node collect torwards exactly one test and then run that test

however that requires refactorings in xdist

**Comment by nicoddemus** [comment]
> So how can I pass an item object to a subprocess?

Just to complement Ronny's answer, the problem is that test items, by design, contain a lot of runtime information which are very hard (impossible) to marshal, for example contains attributes that point to the current fixture manager, config object, etc etc.

**Comment by fontealpina** [comment]
Many thanks for the clarity of the response.
For me this issue can be closed.
